### ISSUE TITLE
New character templates.

### ISSUE THREAD

**Issue by siliconserf** [olivierkes/manuskript#252]
Feature Request: User Editable Templates for Character, World, and Plot entries. 

In the current version, Manuskript allows adding characters to the project. You hit the '+' icon and the code creates the new character object from a hard coded template. What I would like to see is to have a Template section (above Debug) that has user editable versions of addable items. This allows "pre-loading" items like "Characters" with "boilerplate" information. As an example, there are questions that I would like to have show up (optionally) in the Character/Notes section when I add a new character. Snowflake includes a lot of suggested data entries for a character bio. Currently you have to add each of those by hand for each character. Multiple character templates would support different defaults for different types of characters (important, secondary, walk-on, and only referred to).

**Comment by olivierkes** [comment]
Thanks for those ideas.

Right now, I'm refactoring the code, and I want to make characters/word/plot items similar to outline items (editor / text items). This would simplify a lot coding because the same code to search for one type of item could be use for the other.

But this could also allow for more radical UI changes. For example:

![image](https://user-images.githubusercontent.com/1192933/33555359-60f20832-d900-11e7-98b5-d6e13bc7904c.png)

Some possibilities:

  - Every of those items *could* be regrouped in the same view (as in the fake screenshot above)
  - We could use split view to have a parallel view of a text and a character for example.
  - Characters and plots could be within "folders" or "groups" (instead of having one long list categorized only by importance)
  - The metadata pane could be used also for characters
  - …
  - And finally, to answer your initial post, a template system could be created for any of those type of items (I'm thinking of a normal folder that we would create item within, and one we declare this folder as "template folders" any item within can be used as a base for creating a new item)

What do you think? (I'd like also to have @hugo-id and @gedakc's take on that)

**Comment by olivierkes** [comment]
(and also @thoronaug ^^)

**Comment by siliconserf** [comment]
Excellent! That is very much what I had in mind. And I thought the logic would apply nicely to other kinds of objects. Glad to see the efficiency of coding supports this.

**Comment by anicede** [comment]
That tree view is marvellous, @olivierkes, it seems it might save a lot of clicking back and forth.

One thing I've been hungry for is plots in plain text rather than xml as I do a lot of outlining on the go on my tablet. Would this be possible?

**Comment by gedakc** [comment]
Thanks @olivierkes for asking my opinion.

I like the idea of being able to re-use the code for the different areas of editor/character/plot/world.  This should mean more easily maintainable code.  I also like that it would mean we could get rid of the Navigation pane, and hence free up more screen space.

One drawback I foresee is when a large book is expanded to show all chapters (Dan Brown novels > 100 chapters) this would push other items like characters, plots, and worlds off the visible screen.  This would force lots of scrolling to find these other items. With the current 0.6.0 GUI I can quickly switch between editing my novel over to checking/adding a character trait.  Then just as quickly I can return to the editor where I left off.

If there were a way to anchor these other items (chapters, plots, world) at the top of the project tree then that might be one way to address my navigation concern.

**Comment by siliconserf** [comment]
The mock-up looks good. To address the navigation concern caused by a high number of chapters, I suggest a vertical slider control that pops up when the number of lines displayed on the left column exceeds what the size of the window permits. The same horizontally for when item titles get to long for the window. Lastly, make the left column re-sizable much the way the over-all program window behaves. It is the kind of behavior that might be expected from using word processors and web browsers.

Along with the above I suggest a "collapse" icon that affects all items in a level. It would work like this: If you select one of a large number of chapters, hitting the "collapse" control would move focus to the next higher level of organization and collapsing all of the documents under that level. 

I'm a little confused by "...get rid of the navigation...". Are you actually referring to the plot list window that pops up when you select Plots? Because clicking the Navigation tab causes the left hand window to appear and disappear.

**Comment by olivierkes** [comment]
Thanks for your feedbacks!
It's still for the future, but it helps me see where we are going.

>One thing I've been hungry for is plots in plain text rather than xml as I do a lot of outlining on the go on my tablet. Would this be possible?

Sure! I'll have to think of a file format n°2 anyway. It makes sense that plots are in plain text.

>If there were a way to anchor these other items (chapters, plots, world) at the top of the project tree then that might be one way to attempt to address my navigation concern.

That's my main concern with that model. The same problem will happen if there is a large number of characters. Imagine the Game of Thrones series in a manuskript project..! My idea was to add a *filter pannel*: maybe 4 toggle buttons to show/hide item types (texts, plots, characters, world items — we could add a research folder in the future), and a text input to filter items (by title or content).

> Lastly, make the left column re-sizable much the way the over-all program window behaves. It is the kind of behavior that might be expected from using word processors and web browsers.

Yes that's the way it is now and it will remain, it's important!

>Along with the above I suggest a "collapse" icon that affects all items in a level.

Yes that could be useful. Right now, when right-clicking on a item, we can expand/collapse that items, or all. Expand/collapse all is useful, but the single item is not (we can click on the arrow). It could be replaced with "Expand/Collapse siblings".

**Comment by thoronaug** [comment]
Thanks @olivierkes for naming me.

That escalated quickly! XD
Well, I thought this thread was about character templates, and now we're talking about changing UI!

The tree view look awesome, but is it really that useful? I like that view for characters, plots, etc, but if there are so many characters, so many plots, so many chapters it would end in an unclean pane with too much information on in.
Maybe I'm wrong, because I didn't have enough time to examine every feature in Manuskrip but I think the Outline and the Editor (in the ouline view mode) sections are very similar. Perhaps you can merge these two in a bigger Outline, where double clicking an item leads you to the edition of the item and where you can include the tree view for characters, plots, objects, etc.

And keep the Navigation Panel, removing some unused icons. I know @gedakc doesn't like it, and maybe he is right. But if you shink the width of the panel to the icons size, it's pretty small.

This is only my opinion and I respect the others'. Thank you so much for asking.

**Comment by gedakc** [comment]
> And keep the Navigation Panel, removing some unused icons. I know @gedakc doesn't like it,
> and maybe he is right. But if you shink the width of the panel to the icons size, it's pretty small.

@thoronaug I actually prefer the Navigation panel because it allows me to quickly jump between the editors and characters.  Perhaps @olivierkes could keep the navigation panel, yet make some of the project tree code generic so that it could be used for more than just the editor.

**Comment by hugo-id** [comment]
I'm actually an incredibly big fan of the system you already have, Scrivener has this standard display system and I **loathe** it:
![image](https://user-images.githubusercontent.com/30918735/33765851-fdda88a6-dc6e-11e7-99d5-ecf6d3692eec.png)

(Especially when ctrl+tab works as I can simply hide the navigation panel and switch between them seamlessly)

But the general idea of changing the character info stuff would be very welcome, I **love** your character concept and management, but its completely useless for me. Many of the basic fields are awesome for a simple character, but for a more complex character, many of these things are just not applicable (or it just isn't how my mind works)

The four tabs of a character (Basic Info, Summary, Notes, Detailed Info) seem entirely superfluous and I would prefer a character to be a single page. All complex elements in Manuskript (Characters, Plots, And Worlds) have multiple tabs, and honestly I think it's sloppy design. I leave 90% of the fields empty, so how can I justify **three** completely blank tabs for the one thing I'm actually using at the moment.

Basic Info:
Not going to argue with it, but the definition of "basic" does't gel with me. All of these fields seem decidedly un-basic.
Summary:
Why the hell do I need a whole tab for this when there's literally already a provision for this on the first tab (One paragraph summary), which has no size restrictions and I'd trade a scrollbar for an entirely separate tab any day.
Notes:
Very basic functionality, why isn't this a box under "Basic Info"?
Detailed Info:
This was the one I had the greatest hopes for. But it's unsearchable. Restricted to a single line, so then why **must** I have a "Name" & "Value"?
Its an overly-restrictive "Notes" section with no reason for its increased restrictions. Seems like a completely missed opportunity to  implement the [linked value](https://github.com/olivierkes/manuskript/issues/248) and [custom tag](https://github.com/olivierkes/manuskript/issues/241) features, this tab is begging for something to make its existence have a purpose

A character should be **a** tab, like how a scene is at present. [Custom tags](https://github.com/olivierkes/manuskript/issues/241) will allow each author to tune that page to his/her personal flow. Personally, mine would be:

Tab 1: "BASIC INFO"
Title: Searchable + Linkable
First Name: Searchable + Linkable
Last Name: Searchable + Linkable
Nickname: Searchable + Linkable
Faction: Searchable
Notes: Private
References: Private
(Image support would be cool, but in lieu I'll just lump that under references)
Character "Scenes": Mockup I made earlier, [from here](https://github.com/olivierkes/manuskript/issues/248)
![image](https://user-images.githubusercontent.com/30918735/33515484-63b7b254-d7b8-11e7-99e1-3213779c3b44.png)


As I'm sure you'll notice, that's nothing like your basic setup, and is also inapplicable to 90% of people's works, but that's what my hand-written notes look like. I understand Manuskript started off as a program to ease people into using the Snowflake method, but I, and many others, don't use it. So a lot of these seemingly obvious and important choices just seem bizarre (Eg, four levels of story summarisation)

TLDR:
Please don't add another tab for controlling Character templates
Please do allow setting a character's current template via the "Importance" slider at the top. This is a gracious way to insert the advanced functionality and keeping things pretty, it will also neatly allow the templates to be used as a grouping mechanism, just how it is at present.
Please do handle this fancy custom-taggage with simple "Nickname: ..." syntax so that they're still stored in plain txt files
Please do allow the new character templates to specify which fields are searchable (deep search or cheat sheet search) and / or linkable
Please do remove extraneous tabs
Please do ask for further clarification. I've been swamped, so didn't have time to make a pretty mock-up, but am really glad to provide one / give advice when I'm on holidays.
Please don't take offence at potentially harsh criticism
Please do read my feedback in the voice of Sean Connery

Mildly off-topic, but how I could get directly involved?
I spend so much time making mock-ups I find myself asking "wouldn't it be better to just code this directly?"
(Mostly just in UI, don't have time / knowledge to get pulled into backend stuff)

**Comment by olivierkes** [comment]
## Navigation menu

If/when the code is refactored and working and everything, this could easily be switched in settings. 

For the snowflake method, I think I want to add in the future a dock widget (like "Cheat sheet"), that will display the progress in the method, detail the next step, and other useful thing. Right now the process is not very visual. 

## Character tab (and others)
You're right @hugo-id , I think it all calls for simplification.

What's the point of having some field searchable and other not? and some field linkable and other not? (In my proposition below, everything would be searchable and linkable)

>This is a gracious way to insert the advanced functionality and keeping things pretty, it will also neatly allow the templates to be used as a grouping mechanism, just how it is at present.

Not sure I understand that grouping mechanism.

This is an idea I have in mind:

![image](https://user-images.githubusercontent.com/1192933/33774452-8c506a94-dc3b-11e7-9412-992f22796fd2.png)

The character is simple a big text area. Every custom field is define like that: `Name: value`. Value can be multiple lines (Either: new lines start with a space, or new lines are added to previous field if they do not match the syntax). The syntax could be different, it's just an idea.
Note: In the screen shot I used `**Name:**` to make it bold, but in would be syntax highlighted so not needed.

Manuskript parse that text field and extract custom fields, so it is:

- Completely customizable
- Compatible with the snow flake method (if "snowflake mode" is activated, then every character comes populated with empty field), or any other approach
- Easily compatible with templates (a templates is simply a text area with fields names and no values). 
- Easily compatible with in-text references (#248)
- It's easy to save in plain text and to edit in other softwares
- Searching is easy (`"search term" in character.text()`) and easily customizable to be made more powerful (something like `gender:male AND army:slytherin`)

In the screenshot, I used an asterix (`Name*: value`) to signify a transversal field. My idea is that if you type an asterix, when mansukript parses it, it automatically adds the field by default to every other character.

I know this type of approach works very well for me: I love plain text, I think it's more powerful and flexible than anything else.

**Comment by thoronaug** [comment]
I agree with @hugo-id and I don't. Yes, I think a change in the characters section is needed. It need to be simplified IMHO. But @olivierkes idea seems too much simple for me. The way I write (and so many others) implies make the more detailed characters as possible. And this means to have the characters' information very well organized.

I like the idea of a big text field but not for all of the character info. It's perfect for custom fields, anyway, but the **very** basic info, like gender, eyes color, race, history, motivations (I mean common info we should fill in **every** character) must be in a separated tab and in a easy to view format so you can see it at a glance.
I like the simplicity of plain text too, but you have to agree with me it's not the most visual format.

For me, it's at least so important (if not more important) create detailed believable characters than writing the history itself. So simplifying character section is needed but not so much.

**Comment by caseyjp11** [comment]
One of the things I do like (a lot) about scrivener is that a character sheet is a template that can be placed in a folder.
For us "world builder" writers, being able to sort the characters by location/world/setting would make things a LOT simpler in Manuskript, as well as better organizationally.  While I'm not asking for a scrivener "dup", there are some things it is outstanding at doing, and having templates for setting/character sketch/etc. is one of them.

An example here from a current WIP in the linux scrivener beta:
 
![screenshot from 2018-01-12 22-21-00](https://user-images.githubusercontent.com/11329574/34903084-edc66104-f7e6-11e7-8ff0-d5ca1080e62f.png)

**Comment by mrjackphil** [comment]
@olivierkes that's what I'm looking for in this type of sofware!
It pretty useful in game development when you have items/npc/locations and you want to have information about their parameters from one peace of software. 

Custom fields will give flexibility what I'm looking for. Any ideas when it'll be implemented?

**Comment by Zetawilk** [comment]
Is it necessary to have two panes for "Motivation" and "Goal"? My instinct is that this information is very similar, or that motivation can be split between "goal" and "conflict" (in the case of internal conflicts). 

I'm also not certain what "epiphany" refers to. My guess is that it refers to a turning point in the narrative regarding dynamic characters who change over the course of a story, but I can't see how it would be as useful for a static character. 

I also question the placement of the two summaries. "One Sentence Summary" has plenty more space than just one sentence, and seems as though it should be higher up on a character's information, while "One Paragraph Summary" seems general enough to also be listed toward the top. 

Under "Basic Info", I wonder if a pane labeled "Physical Description" would be useful to people. 

I'm curious to know how people use "Summary" versus "Notes". I've been jotting down random snippets of info about a character in their "Notes" tab, and I've been using the "Summary" tab for the biography. 

The "Detailed Info" tab could potentially use a new title, but as of right now, I'm not certain what. While this tab is broken for me (as mentioned elsewhere), I like the idea of a statistical trait page for characters, because it helps me to easily compare and contrast characters' capabilities in relation to one another. A friend of mine on Discord recently pointed out to me when I showed him the program that people planning tabletop campaigns would find an expanded version of this tab terribly useful.

**Comment by siliconserf** [comment]
**Summary:**
Each summary should be an expansion of the previous level. This will make more sense after you've read Dr. Ingermanson's "How to Write a Novel Using the Snowflake Method".  Sentence (Elevator Blurb) -> a 5 sentence Paragraph->5 Paragraphs-> 4-5 pages of paragraphs. The book is entertaining, and many use his approach. Your mileage may vary.

**Characters:**
Changing character importance requires clicking on the slider bar inside the character dialog, just below the character name. Right now you cannot effect that change by dragging the name in the list on the left.

_Basic Info:_
Changing character importance requires clicking on the slider bar below the name field inside the character dialog, just below the character name. Right now you cannot effect that change by dragging the name in the list on the left.

_Motivation_ prompts goals, but are not the same. Your motivation might be to stop being hungry, your goal might be to start a restaurant. And both can be internal and external. And note that in each scene the POV character should have a scene goal separate from the main plot goal. But Minor (or even static) characters, especially walk-ons, may not need to have their basic information filled in--a bellhop walk-on might just be doing his job and need no deep dive into his early childhood and dreams. However, your Protagonist and Antagonist should be looked at closely. Many successful authors refuse to start writing until they understand their primary and secondary characters and are thoroughly familiar with them. But don't make the mistake of filling in too many details - they have a tendency to leak into the story when they might better be left out.

Physical appearances and personal information should be entered in the _Detail Info_ tab. Do it with any characteristic you define in your story, like eye color or hobby, _when such a detail is mention (defined)_. That way if you define it again, when you go to enter the color into the character bible you'll see you had already defined it and can decide how to resolve the conflict. The primary purpose is really to avoid sudden, unexplained changes in physical characteristics that constitute plot holes. But do not define a detail before it is done in a scene unless you really need to see that detail in your mind while working out the plot. It is better to let the reader supply the color in their imagination. It's the principle of having the consumer add an egg to a cake mix.

An epiphany is a breakthrough in thinking. In this case it's the revelation of the real problem the protagonist has, instead of what he thought he had at the beginning. Think the Taco Bell Chihuahua realizing he needs a bigger box to trap Godzilla.

**Templates:** Yeah, I'd like to see the program allow the user to create their own templates. The work-arounds are a) save a "template" project in the _sample-projects_ sub-directory, or b) save someplace in your documents folder tree where you can open it as an "existing project", and immediately save it in a working directory for your new project. BTW - ALWAYS save with the Single File option turned off, and NEVER open more than one instance of _Manuskript_ - the Python interpreter doesn't support that and the results are _interesting_ if you don't have to try to recover from it.
